### PR TITLE
feat(runbook): Add zombie execution runbook

### PR DESCRIPTION
**Pull request by robzienert**

### CODE REVIEW

File: _data/navigation.yml [hunk 1]
```diff
@@ -307,11 +307,6 @@ guides:
             url: /guides/user/canary/judge/
           - title: Best practices
             url: /guides/user/canary/best-practices/
-  - title: Operator How-Tos
```
Comment (tomaslin) [inline _data/navigation.yml:310]: section removed by accident?

File: _data/navigation.yml [hunk 2]
```diff
@@ -307,11 +307,6 @@ guides:
...
-    collapsed: true
-    children:
-      - title: Custom Webhook Stages
-        url: /guides/operator/custom-webhook-stages/
```
Comment (robzienert) [inline _data/navigation.yml:314]: Duplicate nav.

File: guides/runbooks/orca-zombie-executions.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+{% include toc %}
+
+Aliases: `orphaned execution`
+
+A pipeline or task is "stuck" in a `RUNNING` state but not doing anything.
+
+# Diagnosis
+
+Logs will be emitted regularly for Executions that are currently running in Orca.
```
Comment (tomaslin) [inline guides/runbooks/orca-zombie-executions.md]: an example of how this log line looks like describing pipeline id might be useful here

File: guides/runbooks/orca-zombie-executions.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+Found zombie executionType=pipeline application=myapplication executionName=myexample executionId=01CS076X85RX6MWBTQ0VGBF8VX
+```
+
+An alert should be set on the metric `queue.zombies`, triggering whenever there is a count greater than 0.
+
+# Remediation
+
+## Rehydrate Queue
+
+You can attempt to re-hydrate the queue using an [admin API in Orca](https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/admin/web/QueueAdminController.kt#L33), which must be called directly as it is not exposed through Gate.
```
Comment (tomaslin) [inline guides/runbooks/orca-zombie-executions.md]: It's not clear here what rehydrating a queue is or what it does. Would be great if that was defined.

File: guides/runbooks/orca-zombie-executions.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+This command can take either a single execution or operate on all executions within a time range. **This command will dry-run by default.**
```
Comment (tomaslin) [inline guides/runbooks/orca-zombie-executions.md]:
```suggestion
This command can take either a single execution or operate on all executions within a time range. **This command will dry-run by default.** To actually rehydrate the queue pass in the query parameter dryRun=false.
```

File: guides/runbooks/orca-zombie-executions.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+## Canceling the Execution
+
+If the Execution cannot be rehydrated, it will need to be canceled. You can cancel the Execution via the UI.
+
+# Known Causes
+
+A zombie execution is one that has a `status` in the database as `RUNNING` but there are no messages in Orca's work queue or unacked set. 
+In this scenario, Orca will never complete the work. 
+This can happen due to a loss of the Redis instance backing the Orca queue, or in prolonged unreliable networks.
```
Comment (duftler) [inline guides/runbooks/orca-zombie-executions.md]: Can you provide more detail describing how/why losing the Redis instance can cause this situation?

File: _data/navigation.yml (hunk 1 above)
Comment (robzienert) [reply _data/navigation.yml:310]: Nay. The nav section was a duplicate.

File: guides/runbooks/orca-zombie-executions.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+{% include toc %}
+
+Aliases: `orphaned execution`
+
+A pipeline or task is "stuck" in a `RUNNING` state but not doing anything.
+
+# Diagnosis
+
+Logs will be emitted regularly for Executions that are currently running in Orca.
+If no logs have been emitted for over 10 minutes for a `RUNNING` Execution, it is very likely a zombie.
```
Comment (robfletcher) [inline guides/runbooks/orca-zombie-executions.md:17]: Also, looking at the queue itself can be instructive.

File: guides/runbooks/orca-zombie-executions.md (hunk 6 above)
Comment (robzienert) [reply guides/runbooks/orca-zombie-executions.md]: Updated, please take a look.

File: guides/runbooks/orca-zombie-executions.md (hunk 4 above)
Comment (robzienert) [reply guides/runbooks/orca-zombie-executions.md]: Updated. I also moved a couple sentences from the `Known Causes` part up to the summary, which I think sets better context for the rest of the doc. WDYT?

File: guides/runbooks/orca-zombie-executions.md (hunk 7 above)
Comment (robzienert) [reply guides/runbooks/orca-zombie-executions.md:17]: If that's the case, I'd like to add an admin API to inspect the queue. I think guiding people to inspect the internals of the queue with Redis-fu is dangerous (and maybe not even possible depending on the security rules). I'm happy to add more admin functionality--I think we need it--but I'd like some guidance on what explicitly you're checking. Inspection is also highly implementation-specific. If we add a Dynomite queue backend, SQS or Google Pub/Sub, I have strong doubts inspection would be transferrable or even possible.

File: guides/runbooks/orca-zombie-executions.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+An alert should be set on the metric `queue.zombies`, triggering whenever there is a count greater than 0.
+
+# Remediation
+
+## Rehydrate Queue
+
+If the Execution is a zombie, there's no messages on the work queue for that Execution.
+You can attempt to re-hydrate the queue using an [admin API in Orca](https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/admin/web/QueueAdminController.kt#L33), which must be called directly as it is not exposed through Gate.
```
Comment (tomaslin) [inline guides/runbooks/orca-zombie-executions.md]:
```suggestion
You can attempt to re-hydrate the queue --- reissue messages onto the work queue based on the last stored state --- using an [admin API in Orca](https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/admin/web/QueueAdminController.kt#L33), which must be called directly as it is not exposed through Gate.
```

File: guides/runbooks/orca-zombie-executions.md (hunk 7 above)
Comment (tomaslin) [reply guides/runbooks/orca-zombie-executions.md:17]: I like the changes. Reads better.

File: _data/navigation.yml [hunk 9]
```diff
@@ -337,8 +332,10 @@ guides:
         url: /guides/runbooks/api-rate-limiting/
       - title: Sharding Spinnaker
         url: /guides/runbooks/sharding-spinnaker/
-      - title: Orca Quality of Service
+      - title: "Orca: Quality of Service"
```
Comment (dorbin) [inline _data/navigation.yml:335]: @robzienert 
Any reason you added double-quotes  here? Looks like they don't show in the rendered nav, but I don't see them used anywhere else in this file.

File: guides/runbooks/orca-zombie-executions.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  nav: guides
+---
+
+{% include toc %}
+
+Aliases: `orphaned execution`
+
+A pipeline or task is "stuck" in a `RUNNING` state but not doing anything.
+
+A zombie Execution is one that has a `status` in the database as `RUNNING` but there are no messages in Orca's work queue or unacked set.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md]: I think you should swap this paragraph with the sentence above. Actually, delete that sentence and incorporate it in this paragraph:

"A zombie Execution is one that has a `status` in the database of `RUNNING` but there are no messages in Orca's work queue or unacked set&mdash;the pipeline or task is not doing anything."

File: guides/runbooks/orca-zombie-executions.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+This setting is expensive with the `RedisExecutionRepository` and is disabled by default.
+You can enable this detection by setting `queue.zombieCheck.enabled: true` in your Orca configuration.
+
+When enabled, any discovered zombies will be logged out, as well as emitted via a metric:
+
+```
+Found zombie executionType=pipeline application=myapplication executionName=myexample executionId=01CS076X85RX6MWBTQ0VGBF8VX
+```
+
+An alert should be set on the metric `queue.zombies`, triggering whenever there is a count greater than 0.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md]: "...should..."

Another opportunity for some guidance? Maybe: "If you've enabled the zombie check, set an alert on the metric..."

File: guides/runbooks/orca-zombie-executions.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+# Remediation
+
+## Rehydrate the Queue
+
+If the Execution is a zombie, there's no messages on the work queue for that Execution.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md]: s/there's/there are/

File: guides/runbooks/orca-zombie-executions.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+## Cancel the Execution
+
+If the Execution cannot be rehydrated, it will need to be canceled. You can cancel the Execution via the UI.
+
+## Known Causes
+
+Zombie Executions can occur due to a loss of the Redis instance backing the Orca queue, or in prolonged unreliable networks.
+If you're using the `RedisExecutionRepository`, it's likely that when you lose the Redis backing the queue, you're also losing all running Executions.
+However, while using the SQL backend for Orca, losing the Redis means you're only losing the in-flight work queue state, not the state of the pipelines.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md]: Link "using the SQL backend" to /setup/productionize/persistence/orca-sql/?

File: guides/runbooks/orca-zombie-executions.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+If any part of an Execution cannot be re-hydrated, the entire Execution will be skipped.
+
+## Cancel the Execution
+
+If the Execution cannot be rehydrated, it will need to be canceled. You can cancel the Execution via the UI.
+
+## Known Causes
+
+Zombie Executions can occur due to a loss of the Redis instance backing the Orca queue, or in prolonged unreliable networks.
+If you're using the `RedisExecutionRepository`, it's likely that when you lose the Redis backing the queue, you're also losing all running Executions.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md:118]: Can "using the `RedisExecutionRepository`" be linked to anything? Do you know if we describe that anywhere? I don't find any doc for `RedisExecutionRepository`.

File: guides/runbooks/orca-zombie-executions.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+Aliases: `orphaned execution`
+
+A pipeline or task is "stuck" in a `RUNNING` state but not doing anything.
+
+A zombie Execution is one that has a `status` in the database as `RUNNING` but there are no messages in Orca's work queue or unacked set. 
+In this scenario, Orca will never complete the work. 
+
+# Diagnosis
+
+Logs will be emitted regularly for Executions that are currently running in Orca via the `QueueProcessor` class, which will look similar to the following example.
```
Comment (dorbin) [inline guides/runbooks/orca-zombie-executions.md:16]: @robzienert I notice you've got all instances of "Execution/s" capitalized. I *think* it doesn't need to be. But feel free to set me straight.

**Review by dorbin — COMMENTED** [review]
@robzienert 
Just a few comments/questions.

File: _data/navigation.yml (hunk 9 above)
Comment (robzienert) [reply _data/navigation.yml:335]: My editor was coloring the value weird due to the colon in the title.

File: guides/runbooks/orca-zombie-executions.md (hunk 14 above)
Comment (robzienert) [reply guides/runbooks/orca-zombie-executions.md:118]: We don't have any documentation for the RedisExecutionRepository. It's the default if SQL hasn't been configured.

File: guides/runbooks/orca-zombie-executions.md (hunk 15 above)
Comment (robzienert) [reply guides/runbooks/orca-zombie-executions.md:16]: I was capitalizing it to distinguish between the model Execution (representing either pipelines or orchestrations) versus an execution of a pipeline or orchestration. Happy to change it.

File: guides/runbooks/orca-zombie-executions.md (hunk 6 above)
Comment (duftler) [reply guides/runbooks/orca-zombie-executions.md]: LGTM. Thanks.

File: guides/runbooks/orca-zombie-executions.md (hunk 15 above)
Comment (dorbin) [reply guides/runbooks/orca-zombie-executions.md:16]: Nah, it's fine

**Review by dorbin — COMMENTED** [review]
LGTM

**Pull request merged**